src/config/mongoose_config_spec.erl

```diff
@@ -79,23 +85,261 @@ general() ->
        process = fun ?MODULE:process_ctl_access_rule/1,
        format = prepend_key
       }.
 
-process_root(KVs) ->
-    true = lists:any(fun(#local_config{key = hosts}) -> true;
-                        (_) -> false
-                     end, KVs),
-    KVs.
+%% path: listen
+listen() ->
+    Keys = [<<"http">>, <<"c2s">>, <<"s2s">>, <<"service">>],
+    #section{
+       items = maps:from_list([{Key, #list{items = listener(Key),
+                                           format = none}} || Key <- Keys]),
+       format = local_config
+      }.
+
+%% path: listen.*[]
+listener(Type) ->
+    ExtraItems = listener_items(Type),
+    #section{
+       items = ExtraItems#{<<"port">> => #option{type = integer,
+                                                 validate = port},
+                           <<"ip_address">> => #option{type = string,
+                                                       validate = ip_address},
+                           <<"proto">> => #option{type = atom,
+                                                  validate = {enum, [tcp, udp, ws, wss]}},
+                           <<"ip_version">> => #option{type = integer,
+                                                       validate = {enum, [4, 6]},
+                                                       process = fun ?MODULE:process_ip_version/1,
+                                                       format = item}},
+       required = [<<"port">>],
+       process = fun ?MODULE:process_listener/2
+      }.
+
+listener_items(<<"http">>) ->
+    #{<<"tls">> => http_listener_tls(),
+      <<"transport">> => http_transport(),
+      <<"protocol">> => http_protocol(),
+      <<"handlers">> => http_handlers()
+     };
+listener_items(Type) ->
+    ExtraItems = xmpp_listener_items(Type),
+    ExtraItems#{<<"hibernate_after">> => #option{type = integer,
+                                                 validate = non_negative},
+                <<"max_stanza_size">> => #option{type = integer,
+                                                 validate = positive},
+                <<"backlog">> => #option{type = integer,
+                                         validate = non_negative},
+                <<"proxy_protocol">> => #option{type = boolean},
+                <<"num_acceptors">> => #option{type = integer,
+                                               validate = positive,
+                                               format = {kv, acceptors_num}}
+               }.
+
+xmpp_listener_items(<<"c2s">>) ->
+    #{<<"access">> => #option{type = atom,
+                              validate = non_empty},
+      <<"shaper">> => #option{type = atom,
+                              validate = non_empty},
+      <<"xml_socket">> => #option{type = boolean},
+      <<"zlib">> => #option{type = integer,
+                            validate = positive},
+      <<"max_fsm_queue">> => #option{type = integer,
+                                     validate = positive},
```

Comment:
There are so many `#option{type=X,validate=Y}` in schema, that it could be useful to add a helper function, so most of options would fit on a single line. 

```erlang
opt(Type, Validate) ->
    #option{type = Type, validate = Validate}.

listener_items(Type) ->
    ExtraItems = xmpp_listener_items(Type),
    ExtraItems#{<<"hibernate_after">> => opt(integer, non_negative),
                <<"max_stanza_size">> => opt(integer, positive),
                <<"backlog">> => opt(integer, non_negative),
                <<"proxy_protocol">> => #option{type => boolean},
                <<"num_acceptors">> => #option{type = integer,
                                               validate = positive,
                                               format = {kv, acceptors_num}}
               }.
```

(but I don't think we would need opt/1, opt/3 and opt/4 functions. Maaybe we can have opt/1).

But on another hand the current format is more consistent... I don't know, it just a comment :D

Reply:
It's worth considering, but I wouldn't change it atm.